### DISCUSSION TITLE
"mpfs-rpmsg-bm"demo bulid with "undefined reference to `sprintf' "error

### DISCUSSION THREAD

**Discussion by EricZhangjie** [polarfire-soc/polarfire-soc-documentation#132] [Applications and demos] [answered]
hello, I use polarfire-soc-amp-examples\polarfire-soc-amp-examples-2022.02\mpfs-rpmsg-bm as a demo, and add some lines that call sprintf function as below
![sprintf](https://user-images.githubusercontent.com/109322517/179128776-94e6508e-8eb6-4164-a318-c2229b3a39e5.JPG)
I add line 11/12/19/20/40, then build it, it report error: undefined reference to `sprintf'.I try it in SoftConsole-v2021.1 and SoftConsole-v2021.3-7.0.0.599.  I press ctrl button and click line 11, it can jump to stdio.h file, and I can find sprintf function declaration in the file.
I try other demos,  polarfire-soc-bare-metal-examples-2021.11\driver-examples\mss\mpfs-hal\mpfs-hal-simple-demo, and polarfire-soc-bare-metal-examples\polarfire-soc-bare-metal-examples-main\driver-examples\mss\mss-mmuart\mpfs-mmuart-interrupt, there is no this problem.
please help me, thank you!

**Comment by AntonKrug** [comment]
You can see declaration (just statement that something somewhere implementes it), but that's not definition (the real implementation).
You can remember like it's in the office scene: https://www.youtube.com/watch?v=EuZeff2y32M
Declaration on its own without a definition will not do anything.

If you tell the linker to not use the default libraries, then they will not be defined. 

There are multiple ways to address the missing printf implementation, for example:

- Change the linker settings (just note there might be a very good reasons why the project doesn't inlcude the default libs):
`project properties -> c/c++ build -> settings -> tool settings -> gnu risc-v cross linker -> general -> "Do not use default libraries (-nodefaultlibs)"`

- Or use/provide a different implementation of the printf, for example:
https://mi-v-ecosystem.github.io/SoftConsole-Documentation/SoftConsole-v2021.3/troubleshooting/project.html?highlight=printf

**Comment by vfalanis** [comment, marked as answer]
Hi @EricZhangjie ,

The mpfs-rpmsg demos in the polarfire-soc-amp-examples directory are compiled in standalone mode (no stdlib), that's the reason why there's no standard implementation of sprintf. However there is an ee_vsprintf function included in the project which can be used for the same purpose.

Example:
ee_vsprintf(info_string, "timer_cnt = %d, \r\n",timer_cnt);

There are also some additional helper functions which will write the string directly to UART after formatting. For example: 

https://github.com/polarfire-soc/polarfire-soc-amp-examples/blob/main/mpfs-rpmsg-freertos/src/application/inc/pingpong_demo.c#L157

For more information please refer to the ee_vsprintf implementation within the project:
https://github.com/polarfire-soc/polarfire-soc-amp-examples/blob/main/mpfs-rpmsg-freertos/src/middleware/misc/ee_vsprintf.c#L605

**Reply by EricZhangjie** [reply]
thanks, I followed the pingpong_demo.c, and used ee_printf function. in this AMP mode demo, how can I compile with stdlib? In project properties -> c/c++ build -> settings, there is no tool settings for users to change the linker settings.
![no_tool_settings](https://user-images.githubusercontent.com/109322517/179346303-c3b43eef-f967-474e-90c1-028e9b2fa527.png)

**Reply by vfalanis** [reply]
stdlib can be enabled by removing the  line "-nostdlib" and "-nodefaultlibs" in the mpfs-rpmsg-bm/Makefile (line 93). After removing those flags, line 93 should looks as follows:

```
$(CMD_PREFIX)$(CC) -T $(LINKER_SCRIPT) -static -nostartfiles -Wl,-Map=$(BINDIR)/output.map -Wl,--gc-sections -o $(BINDIR)/$@ $(OBJS)
```

**Reply by AntonKrug** [reply]
Apologies for confusing you, SoftConsole supports 2 types of projects, Managed and Makefile. Managed projects have GUI with lots of options/settings which then generates a Makefile on demand for you behind the scenes and then build the project. While the Makefile projects have human made hand-crafted Makefiles and SoftConsole just builds them (there is no GUI for changing its settings and you have to edit the Makefile by hand). The instructions I gave in the first comment was for a managed project with GUI settings without realizing myself that you are using the Makefile based project in this case. So take and use @vfalanis instructions over mine.
